### PR TITLE
TEP-0076 (array results and indexing into array) promoted to beta 🎓

### PR DESCRIPTION
**Pull request by pritidesai**
# Changes

[Pipelines 0.38](https://github.com/tektoncd/pipeline/releases/tag/v0.38.0) introduced two alpha features - producing array results and indexing into array param (proposed in [TEP-0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md)). This commit is promoting these two features to `beta` such that these features can be used by the task authors and pipeline authors in a cluster when `enable-api-fields` is either set to `beta` or `alpha`.

These features are still not available in stable API i.e. these features will not work when `enable-api-fields` set to `stable`

Before we can merge this PR, I have following three PRs to update the doc and examples. Please review the outstanding PRs first. **These PRs are merged now.**

- PR #6055 
- PR #6060 
- PR #6092 

Fixes part of https://github.com/tektoncd/pipeline/issues/5688.

Signed-off-by: pritidesai <pdesai@us.ibm.com>

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
TEP-0076 promoted to beta - Array results and indexing into an array is now possible with enable-api-fields set to beta.
```

### CODE REVIEW

**Comment by pritidesai** [comment]
@afrittoli @JeromeJu @Yongxuanzhang I appreciate if you can review the changes in this PR, my team is looking for this feature as part of 0.45. Thanks!

File: pkg/apis/config/feature_flags_test.go
Comment (JeromeJu) [inline pkg/apis/config/feature_flags_test.go:291]: Thanks for the PR @pritidesai this looks good.

Just one NIT question, curious why we would explicitly check this is not Stable instead of testing if the intended flags are set correctly eg. verify stable for `config.EnableStableAPIFields(ctx)`?

**Review by JeromeJu — APPROVED** [review]

**Review by Yongxuanzhang — COMMENTED** [review]
/lgtm
Thanks! Sorry I thought we want to merge other issues fixes before this one. I think it's ok to get this merged in v0.45

File: pkg/apis/config/feature_flags_test.go
Comment (pritidesai) [reply pkg/apis/config/feature_flags_test.go:291]: thank you @JeromeJu for the review 👍 

I am not sure if I followed your question 😜 

This is a test function to validate `CheckAlphaOrBetaAPIFields` i.e. return true if `enable-api-fields` set to either `alpha` or `beta` and false otherwise. The test cases below include testing for all three flags.

Comment (JeromeJu) [reply pkg/apis/config/feature_flags_test.go:291]: I do agree that for `CheckAlphaOrBetaAPIFields` this makes sense and covers all the api feature flags. 

I was curious at first sight about the naming for `CheckAlphaOrBetaAPIFields` but now I think this is fine as an intermediate stage to beta.

**Review by vdemeester — APPROVED** [review]

**Comment by vdemeester** [comment]
Cc @abayer @afrittoli @jerop to remove the hold 😛 but yay 🎉

File: docs/pipelines.md
Comment (afrittoli) [inline docs/pipelines.md:945]: NIT: shouldn't we say that "array results are a beta feature. They can be enabled in the `v1` API by setting the `enable-api-fields` to `beta`

File: docs/tasks.md
Comment (afrittoli) [inline docs/tasks.md:836]: Ditto: shouldn't me mention that arrays are a beta feature?

File: pkg/apis/config/feature_flags_test.go
Comment (afrittoli) [inline pkg/apis/config/feature_flags_test.go:308]: NIT: stable is the default (and will always be), so this feels a bit redundant.

File: pkg/apis/pipeline/v1/result_validation.go
Comment (afrittoli) [inline pkg/apis/pipeline/v1/result_validation.go:52]: NIT: I find it a bit brittle to rely on the sum of case condition to provide full coverage of all cases, I think it would be preferable to do something like:

```go
	default:
            if tr.Type != ResultsTypeString:
               return apis.ErrInvalidValue(tr.Type, "type", "type must be string")
```

**Review by afrittoli — COMMENTED** [review]
Thanks @pritidesai 🎉 - this looks mostly good to me, only a couple of minor things - it would be nice to have the docs updated, especially given the fact that we don't have v1/v1beta separate docs, so the docs should default to the latest API which is v1

File: docs/pipelines.md
Comment (pritidesai) [reply docs/pipelines.md:945]: I am not sure, with `v1beta1` API, for this feature to work, `enable-api-fields` does not have to be `alpha` anymore which means it can be set to either `beta` or `stable`. 

@chitrangpatel how did you approach this when we promoted propagated params/workspaces  features? 🤔 Or these features were not classified as `beta`? will appreciate any guidance. Thanks!

**Comment by abayer** [comment]
/approve

Good by me!

File: docs/pipelines.md
Comment (chitrangpatel) [reply docs/pipelines.md:945]: Propagated Params was moved straight to stable since stable and beta were same at the time.
Propagated workspaces is currently beta meaning that it will work if `enable-api-fields` is `alpha` or `beta` but **not** `stable`.

File: pkg/apis/pipeline/v1/result_validation.go
Comment (pritidesai) [reply pkg/apis/pipeline/v1/result_validation.go:52]: thank you @afrittoli I am not able to follow the concern here, sorry 🙃 😕

Comment (pritidesai) [reply pkg/apis/pipeline/v1/result_validation.go:52]: the `default` is always guaranteed to execute, not sure if it make sense to introduce default with a `condition` 🤔 v/s its better to be explicit with the `case` statement ...

File: pkg/apis/config/feature_flags_test.go
Comment (pritidesai) [reply pkg/apis/config/feature_flags_test.go:308]: Yup I completely agree, `stable` is a default. its just added to be consistent with the 

> /approve
> 
> Good by me!

thanks @abayer 🙏 

@vdemeester can we cancel the hold? I do not think there is anything pending here now. If there is a concern of array params in a `matrix` and `when` expressions, we had examples and doc updated for both features in PR #6092 and PR #6077. @Yongxuanzhang please confirm!

File: pkg/apis/pipeline/v1beta1/result_validation_test.go
Comment (pritidesai) [reply pkg/apis/pipeline/v1beta1/result_validation_test.go:86]: yup completely agree, will create a separate PR. There are many other such cases in the same test file. Thanks!

File: docs/pipelines.md
Comment (jerop) [reply docs/pipelines.md:945]: yea, we should clarify that array results is an beta feature

@pritidesai it seems that in both v1 and v1beta1, we check that beta feature flag is enabled -- @lbernick we may need clarification in the TEP about whether the flag should be the same in both versions

File: pkg/apis/config/feature_flags.go
Comment (jerop) [inline pkg/apis/config/feature_flags.go:350]: not related to this PR -- @chitrangpatel what do you think of cleaning up propagated workspaces to reuse this function? just to make the logic clearer?

Comment (chitrangpatel) [reply pkg/apis/config/feature_flags.go:350]: Yes, can do! I think it's best to stay consistent. Once this is merged, I will open an issue and implement a fix.

File: docs/pipelines.md
Comment (pritidesai) [reply docs/pipelines.md:945]: > Propagated workspaces is currently beta meaning that it will work if enable-api-fields is alpha or beta but not stable.

Thanks @chitrangpatel for getting back. How did we document that propagated workspace is `beta` feature? At the same time, I am checking docs. I just want to make sure its consistent across many different features.

File: pkg/apis/config/feature_flags.go
Comment (jerop) [reply pkg/apis/config/feature_flags.go:350]: sounds good, thanks @chitrangpatel!

File: docs/pipelines.md
Comment (chitrangpatel) [reply docs/pipelines.md:945]: https://github.com/tektoncd/pipeline/pull/6030/files is all I had done.

Comment (pritidesai) [reply docs/pipelines.md:945]: So I just checked the PR #6030, there is no specific guidelines other than the replacing `alpha` with `beta` in the docs 🙃 I will update the docs here but it's really very confusing for folks using `v1beta1` APIs.

**Comment by pritidesai** [comment]
thank you @Yongxuanzhang @afrittoli @vdemeester @abayer @JeromeJu @jerop and @chitrangpatel 🙏 

I have updated this PR to document that this feature is `beta` feature. I am looking for lgtm unless there is anything else pending. Thanks a bunch!

**Review by jerop — APPROVED** [review]

**Comment by Yongxuanzhang** [comment]
/lgtm
Thanks!

File: pkg/apis/pipeline/v1/result_validation.go
Comment (afrittoli) [reply pkg/apis/pipeline/v1/result_validation.go:52]: The `default` is not always guaranteed to execute, it won't be executed if any of the options before matches. 
But it guarantees execution if the value was not caught by any of the `case` statements. 
My proposed alternative is not very different indeed, I think a better one would be to explicitly succeed on "string" and fail with a different error message in the default case.
Not blocking for this PR anyways, I can make a small update once this is merged.

**Pull request merged**
